File: AngularJS/src/org/angular2/codeInsight/Angular2SpecificHandlersFactory.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.codeInsight;

import com.intellij.lang.javascript.DialectDetector;
import com.intellij.lang.javascript.JavaScriptSpecificHandlersFactory;
import com.intellij.lang.javascript.ecmascript6.TypeScriptQualifiedItemProcessor;
import com.intellij.lang.javascript.findUsages.JSDialectSpecificReadWriteAccessDetector;
import com.intellij.lang.javascript.psi.ecmal4.JSClass;
import com.intellij.lang.javascript.psi.impl.JSReferenceExpressionImpl;
import com.intellij.lang.javascript.psi.resolve.*;
import com.intellij.lang.typescript.resolve.TypeScriptTypeHelper;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.resolve.ResolveCache;
import org.angular2.codeInsight.refs.Angular2ReferenceExpressionResolver;
import org.angular2.entities.Angular2ComponentLocator;
import org.angular2.findUsages.Angular2ReadWriteAccessDetector;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class Angular2SpecificHandlersFactory extends JavaScriptSpecificHandlersFactory {

  @Override
  public @NotNull ResolveCache.PolyVariantResolver<JSReferenceExpressionImpl> createReferenceExpressionResolver(
    JSReferenceExpressionImpl referenceExpression, boolean ignorePerformanceLimits) {
    return new Angular2ReferenceExpressionResolver(referenceExpression, ignorePerformanceLimits);
  }

  @Override
  public <T extends ResultSink> QualifiedItemProcessor<T> createQualifiedItemProcessor(@NotNull T sink, @NotNull PsiElement place) {
    JSClass clazz = Angular2ComponentLocator.findComponentClass(place);
    if (clazz != null && DialectDetector.isTypeScript(clazz)) {
      return new TypeScriptQualifiedItemProcessor<>(sink, place.getContainingFile());
    }

    return super.createQualifiedItemProcessor(sink, place);
  }

  @Override
  public @NotNull JSImportHandler getImportHandler() {
    return new Angular2ImportHandler();
  }

  @Override
  public @NotNull JSTypeEvaluator newTypeEvaluator(@NotNull JSEvaluateContext context) {
    return new Angular2TypeEvaluator(context);
  }

  @Override
  public @NotNull AccessibilityProcessingHandler createAccessibilityProcessingHandler(@Nullable PsiElement place, boolean skipNsResolving) {
    return new Angular2AccessibilityProcessingHandler(place);
  }

  @Override
  public @NotNull JSDialectSpecificReadWriteAccessDetector getReadWriteAccessDetector() {
    return Angular2ReadWriteAccessDetector.INSTANCE;
  }

  @Override
  public @NotNull JSTypeGuardEvaluator getTypeGuardEvaluator() {
    return Angular2TypeGuardEvaluator.INSTANCE;
  }

  @Override
  public @NotNull JSTypeHelper getTypeHelper() {
    return TypeScriptTypeHelper.getInstance();
  }
}

File: AngularJS/src/org/angular2/codeInsight/Angular2SpecificHandlersFactory.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.codeInsight

import com.intellij.codeInsight.controlflow.ControlFlow
import com.intellij.lang.javascript.DialectDetector
import com.intellij.lang.javascript.JavaScriptSpecificHandlersFactory
import com.intellij.lang.javascript.ecmascript6.TypeScriptQualifiedItemProcessor
import com.intellij.lang.javascript.findUsages.JSDialectSpecificReadWriteAccessDetector
import com.intellij.lang.javascript.psi.JSControlFlowScope
import com.intellij.lang.javascript.psi.JSElement
import com.intellij.lang.javascript.psi.JSFile
import com.intellij.lang.javascript.psi.ecmal4.JSClass
import com.intellij.lang.javascript.psi.impl.JSReferenceExpressionImpl
import com.intellij.lang.javascript.psi.resolve.*
import com.intellij.lang.javascript.psi.types.guard.TypeScriptTypeGuard
import com.intellij.lang.javascript.psi.util.JSStubBasedPsiTreeUtil
import com.intellij.lang.javascript.psi.util.JSStubBasedScopeHandler
import com.intellij.lang.typescript.resolve.TypeScriptTypeHelper
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.impl.source.resolve.ResolveCache.PolyVariantResolver
import org.angular2.codeInsight.controlflow.Angular2ControlFlowBuilder
import org.angular2.codeInsight.refs.Angular2ReferenceExpressionResolver
import org.angular2.entities.Angular2ComponentLocator
import org.angular2.findUsages.Angular2ReadWriteAccessDetector

class Angular2SpecificHandlersFactory : JavaScriptSpecificHandlersFactory() {
  override fun createReferenceExpressionResolver(
    referenceExpression: JSReferenceExpressionImpl, ignorePerformanceLimits: Boolean): PolyVariantResolver<JSReferenceExpressionImpl> {
    return Angular2ReferenceExpressionResolver(referenceExpression, ignorePerformanceLimits)
  }

  override fun getControlFlow(scope: JSControlFlowScope): ControlFlow {
    return Angular2ControlFlowBuilder().buildControlFlow(scope)
  }

  override fun <T : ResultSink> createQualifiedItemProcessor(sink: T, place: PsiElement): QualifiedItemProcessor<T> {
    val clazz: JSClass? = Angular2ComponentLocator.findComponentClass(place)
    return if (clazz != null && DialectDetector.isTypeScript(clazz)) {
      TypeScriptQualifiedItemProcessor(sink, place.containingFile)
    }
    else super.createQualifiedItemProcessor(sink, place)
  }

  override fun getImportHandler(): JSImportHandler {
    return Angular2ImportHandler()
  }

  override fun newTypeEvaluator(context: JSEvaluateContext): JSTypeEvaluator {
    return Angular2TypeEvaluator(context)
  }

  override fun createAccessibilityProcessingHandler(place: PsiElement?, skipNsResolving: Boolean): AccessibilityProcessingHandler {
    return Angular2AccessibilityProcessingHandler(place)
  }

  override fun getReadWriteAccessDetector(): JSDialectSpecificReadWriteAccessDetector {
    return Angular2ReadWriteAccessDetector
  }

  override fun getTypeGuardEvaluator(): JSTypeGuardEvaluator {
    return Angular2TypeGuardEvaluator
  }

  override fun createTypeGuard(element: PsiElement): TypeScriptTypeGuard {
    return Angular2TypeGuard(element)
  }

  override fun getTypeHelper(): JSTypeHelper {
    return TypeScriptTypeHelper.getInstance()
  }

  override fun getExportScope(element: PsiElement): JSElement? =
    if (element is PsiFile)
      null
    else
      Angular2ComponentLocator.findComponentClass(element)
        ?.containingFile as? JSFile
      ?: super.getExportScope(element)

  override fun resolveReferenceLocallyForAnalysis(reference: PsiPolyVariantReference, referenceName: String): PsiElement? {
    return super.resolveReferenceLocallyForAnalysis(reference, referenceName)
           ?: Angular2ComponentLocator.findComponentClass(reference.element)?.let {
             JSStubBasedPsiTreeUtil.resolveLocally(referenceName, it, false)
           }
  }

}